README.md
PHP-PayPal-IPN
==============

A PHP class for handling Instant Payment Notifications (IPN) from PayPal.

Features:

* Switch between live and sandbox by setting the `use_sandbox` property.
* Supports HTTPS (SSL) transactions by default (HTTP is discontinued by PayPal as of 2016).
* Supports both cURL and fsockopen network libraries by setting the `use_curl` property (cURL is recommended).
* Verifies an HTTP 200 response status code from the PayPal server.
* Get detailed plain text reports of the entire IPN using the `getTextReport()` method for use in emails and logs to administrators.
* Throws various exceptions to differentiate between common errors in code or server configuration versus invalid IPN responses.

Requirements:

* PHP 5.3.0 or newer.
* cURL extension (if `use_curl` property is enabled).

Documentation
-------------

The source code contains PHPDoc style comments which essentially provide a function reference.

Example scripts ranging from basic IPN verification to a complete order processing chain, and a sample report as returned by the `getTextReport()` method, can be found in the [Examples](examples/) folder.

This package is available via [Composer](https://getcomposer.org/) at [dezlov/php-paypal-ipn](https://packagist.org/packages/dezlov/php-paypal-ipn).

### Getting Started

You should understand how the IPN process works conceptually and you should understand when and why you would be using IPN. Reading the [PayPal Instant Payment Notification Guide][1] is a good place to start.

You should also have a [PayPal Sandbox Account][2] with a test buyer account and a test seller account. When logged into your sandbox account there is an IPN simulator under the "Test Tools" menu which you can use to test your IPN listener.

Once you have your sandbox account setup, you simply create a PHP script that will be your IPN listener. In that script, use the `IpnListener()` class as demonstrated in the [Examples](examples/).

Also consider reading this [in-depth tutorial][3] about the IPN process and how to implementing an IPN listener script in PHP.

[1]: https://developer.paypal.com/webapps/developer/docs/classic/products/instant-payment-notification/
[2]: https://developer.paypal.com
[3]: http://www.micahcarrick.com/paypal-ipn-with-php.html

History
-------

Changes between versions are documented in the [CHANGELOG.md](CHANGELOG.md).

This library was forked from [WadeShuler/PHP-PayPal-IPN](https://github.com/WadeShuler/PHP-PayPal-IPN) which in turn was forked from [Quixotix/PHP-PayPal-IPN](https://github.com/Quixotix/PHP-PayPal-IPN).

Below is a brief summary of their differences and reasons for forking:

1. Quixotix's version is no longer maintained and contains various critical issues.
2. WadeShuler's version was meant to address all issues of the original version, however, several breaking changes have been made unnecessarily.
3. The goal of this version is to keep the library operational by fixing any issues, while trying to retain backward compatibility with the original version.

License
-------

This work is licensed under [GPL 2.0](http://choosealicense.com/licenses/gpl-2.0/).

Copyright notice from the original author: [LICENSE](LICENSE).

CHANGELOG.md
# Change Log

This change log contains the highlights of major changes. For details on all finer changes please check the commits history.

## v3.0.0

- Added a sample report.
  - As returned by the `getTextReport()` method.
- Added 5 example scripts.
  - Ranging from basic IPN verification to a complete order processing chain.
- Added user defined path to Certificate Authority (CA) bundle file.
- Reformatted output of `getTextReport()` and added a list of used options. 
- Added `getData()` and `getRawData()` methods.
  - Deprecated `getPostData()` and `getRawPostData()` methods in favor of the new methods.
- Fixed data mishandling issues in `processIpn()`.
- Code reformatting.
- Fixed "*Undefined index: REQUEST_METHOD*" when calling `requirePostMethod()` from CLI.
- Make POST method requirement optional via `requirePostMethod` property.
  - Ignore POST method requirement when data is supplied via a parameter.
- Added `tryProcessIpn()` for processing IPN requests without having to handle exceptions.
  - Internally it calls `processIpn()` but catches any exceptions.
- Restored original error handling in `processIpn()`.
  - It once again throws an exception when an error occurs, like in the original version.
- Fixed "*invalid argument supplied for foreach()*" in `getTextReport()`.
- Rebranding for a new fork.
  - Changed namespace to `dezlov\PayPal`. 

## v2.5.2

Forked from [WadeShuler/PHP-PayPal-IPN](https://github.com/WadeShuler/PHP-PayPal-IPN).

## v2.1.0

Forked from [Quixotix/PHP-PayPal-IPN](https://github.com/Quixotix/PHP-PayPal-IPN).


examples/ipn_local_test.php
<?php

/*
 * An example script for testing pre-made IPN requests locally.
 * 
 * Because we manually provide the data that is normally sent by PayPal,
 * we can run this script from command line or open it in a web browser
 * and see the output.
 */

include(dirname(dirname(__FILE__)).'/src/IpnListener.php');

use dezlov\PayPal\IpnListener;

$listener = new IpnListener();
$listener->use_sandbox = true;

// Example IPN request data, we expect PayPal to tell us that it is "INVALID".
$example_request_data = 'mc_gross=19.95&protection_eligibility=Eligible&address_status=confirmed&payer_id=LPLWNMTBWMFAY&tax=0.00&address_street=1+Main+St&payment_date=20%3A12%3A59+Jan+13%2C+2009+PST&payment_status=Completed&charset=windows-1252&address_zip=95131&first_name=Test&mc_fee=0.88&address_country_code=US&address_name=Test+User&notify_version=2.6&custom=&payer_status=verified&address_country=United+States&address_city=San+Jose&quantity=1&verify_sign=AtkOfCXbDm2hu0ZELryHFjY-Vb7PAUvS6nMXgysbElEn9v-1XcmSoGtf&payer_email=gpmac_1231902590_per%40paypal.com&txn_id=61E67681CH3238416&payment_type=instant&last_name=User&address_state=CA&receiver_email=gpmac_1231902686_biz%40paypal.com&payment_fee=0.88&receiver_id=S8XGHLYDW9T3S&txn_type=express_checkout&item_name=&mc_currency=USD&item_number=&residence_country=US&test_ipn=1&handling_amount=0.00&transaction_subject=&payment_gross=19.95&shipping=0.00';

// Verify IPN request trhough PayPal
$error = null;
$verified = $listener->tryProcessIpn($example_request_data, $error);
if ($verified)
	echo 'IPN verified successfully.'.PHP_EOL;
else
	echo 'IPN error: '.$error.PHP_EOL;

// Print out debug information
print_r($listener->getData());
echo $listener->getTextReport();


src/IpnListener.php
<?php

namespace dezlov\PayPal;

use Exception;

/**
 *  A class for handling Instant Payment Notifications (IPN) from PayPal.
 *
 *  @package    PHP-PayPal-IPN
 *  @link       https://github.com/dezlov/PHP-PayPal-IPN
 *  @author     Denis Kozlov
 *  @license    http://choosealicense.com/licenses/gpl-2.0/
 */
class IpnListener
{
	/**
	 *  If true, the recommended cURL PHP library is used to send the post back
	 *  to PayPal. If flase then fsockopen() is used. Default true.
	 *
	 *  @var boolean
	 */
	public $use_curl = true;

	/**
	 *  If true, cURL will use the CURLOPT_FOLLOWLOCATION to follow any
	 *  "Location: ..." headers in the response.
	 *
	 *  @var boolean
	 */
	public $follow_location = false;

	/**
	 *  If true, the paypal sandbox URI www.sandbox.paypal.com is used for the
	 *  post back. If false, the live URI www.paypal.com is used. Default false.
	 *
	 *  @var boolean
	 */
	public $use_sandbox = false;

	/**
	 *  The amount of time, in seconds, to wait for the PayPal server to respond
	 *  before timing out. Default 30 seconds.
	 *
	 *  @var int
	 */
	public $timeout = 30;

	/**
	 * If true, enable SSL certificate validation of PayPal server when using cURL.
	 *
	 * @var boolean
	 */
	public $verify_ssl = true;

	/**
	 * Require that POST method is used when processing an IPN request from POST data.
	 * Exception is thrown if this check has failed. Default is TRUE.
	 *
	 * @var boolean
	 */
	public $requirePostMethod = true;

	/**
	 * User defined path to Certificate Authority (CA) bundle file. It is used
	 * when SSL certificate validation is enabled via <code>$verify_ssl</code>.
	 * If NULL, default path is used as defined by <code>CERT_BUNDLE_DEFAULT</code>.
	 *
	 * @var string|null
	 */
	public $certificateBundlePath = null;

	private $postData = array();
	private $rawPostData = null;
	private $post_uri = '';
	private $response_status = '';
	private $response = '';

	const PAYPAL_HOST = 'www.paypal.com';
	const SANDBOX_HOST = 'www.sandbox.paypal.com';
	const TARGET_URI_PATH = '/cgi-bin/webscr';

	/**
	 * Default path to Certificate Authority (CA) bundle file.
	 * Path is relative to the root of this library.
	 */
	const CERT_BUNDLE_DEFAULT = '/cert/api_cert_chain.crt';

	/**
	 * Get path to Certificate Authority (CA) bundle file.
	 * If user supplied path is not defined, uses <code>CERT_BUNDLE_DEFAULT</code>.
	 * 
	 * @return string
	 */
	protected function getCertificateBundlePath()
	{
		if ($this->certificateBundlePath !== null)
			return $this->certificateBundlePath;
		else
			return dirname(dirname(__FILE__)).self::CERT_BUNDLE_DEFAULT;
	}

	/**
	 *  Post Back Using cURL
	 *
	 *  Sends the post back to PayPal using the cURL library. Called by
	 *  the processIpn() method if the use_curl property is true. Throws an
	 *  exception if the post fails. Populates the response, response_status,
	 *  and post_uri properties on success.
	 *
	 *  @todo add URL param so function is more dynamic
	 *
	 *  @param  string  The post data as a URL encoded string
	 */
	protected function curlPost($encoded_data)
	{
		$host = $this->getTargetHost();
		$path = self::TARGET_URI_PATH;
		$url = 'https://'.$host.$path;
		$this->post_uri = $url;

		$ch = curl_init();

		if ($this->verify_ssl)
		{
			curl_setopt($ch, CURLOPT_SSL_VERIFYPEER, true);
			curl_setopt($ch, CURLOPT_SSL_VERIFYHOST, 2);
			curl_setopt($ch, CURLOPT_CAINFO, $this->getCertificateBundlePath());
		}

		curl_setopt($ch, CURLOPT_URL, $url);
		curl_setopt($ch, CURLOPT_POST, true);
		curl_setopt($ch, CURLOPT_POSTFIELDS, $encoded_data);
		curl_setopt($ch, CURLOPT_FOLLOWLOCATION, $this->follow_location);
		curl_setopt($ch, CURLOPT_TIMEOUT, $this->timeout);
		curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
		curl_setopt($ch, CURLOPT_HEADER, true);

		$this->response = curl_exec($ch);
		$this->response_status = strval(curl_getinfo($ch, CURLINFO_HTTP_CODE));

		if (($this->response === false) || ($this->response_status == '0'))
		{
			$errno = curl_errno($ch);
			$errstr = curl_error($ch);
			throw new Exception("cURL error: [$errno] $errstr");
		}

		return $this->response;
	}

	/**
	 *  Post Back Using fsockopen()
	 *
	 *  Sends the post back to PayPal using the fsockopen() function. Called by
	 *  the processIpn() method if the use_curl property is false. Throws an
	 *  exception if the post fails. Populates the response, response_status,
	 *  and post_uri properties on success.
	 *
	 *  @todo add URL param so function is more dynamic
	 *
	 *  @param  string  The post data as a URL encoded string
	 */
	protected function fsockPost($encoded_data)
	{
		$host = $this->getTargetHost();
		$path = self::TARGET_URI_PATH;
		$fsockhost = 'ssl://'.$host;
		$port = '443';
		$this->post_uri = $fsockhost.':'.$port.$path;

		$fp = fsockopen($fsockhost, $port, $errno, $errstr, $this->timeout);

		if (!$fp)
		{
			// fsockopen error
			throw new Exception("fsockopen error: [$errno] $errstr");
		}

		$header = "POST ".$path." HTTP/1.1\r\n";
		$header .= "Host: ".$host."\r\n";
		$header .= "Content-Type: application/x-www-form-urlencoded\r\n";
		$header .= "Content-Length: ".strlen($encoded_data)."\r\n";
		$header .= "Connection: Close\r\n\r\n";

		fputs($fp, $header.$encoded_data."\r\n\r\n");

		while(!feof($fp))
		{
			if (empty($this->response))
			{
				// extract HTTP status from first line
				$this->response .= $status = fgets($fp, 1024);
				$this->response_status = trim(substr($status, 9, 4));
			}
			else
			{
				$this->response .= fgets($fp, 1024);
			}
		}

		fclose($fp);
		return $this->response;
	}

	private function getTargetHost()
	{
		return ($this->use_sandbox) ? self::SANDBOX_HOST : self::PAYPAL_HOST;
	}

	/**
	 * Get a value of a variable from processed IPN data. If variable name is
	 * not specified, the full data set is returned as an associative array.
	 * @param string $variable Variable name. If NULL, full data set is returned.
	 *   Default is NULL.
	 * @param mixed $default Default value to be returned if variable does not exist.
	 *   Default is NULL.
	 * @return string|array
	 */
	public function getData($variable=null, $default=null)
	{
		if ($variable === null)
			return $this->postData;
		else if (isset($this->postData[$variable]))
			return $this->postData[$variable];
		else
			return $default;
	}

	/**
	 * Get raw data from processed IPN.
	 * Can be useful for debugging purposes.
	 * @return string
	 */
	public function getRawData()
	{
		return $this->rawPostData;
	}

	/**
	 * Get parsed data from processed IPN.
	 * @deprecated Deprecated, use <code>getData()</code> instead.
	 * @return array
	 */
	public function getPostData()
	{
		return $this->getData();
	}

	/**
	 * Get raw data from processed IPN.
	 * @deprecated Deprecated, use <code>getRawData()</code> instead.
	 * @return string
	 */
	public function getRawPostData()
	{
		return $this->getRawData();
	}

	/**
	 *  Get POST URI
	 *
	 *  Returns the URI that was used to send the post back to PayPal.
	 *  This can be useful for troubleshooting connection problems.
	 *
	 *  @return string
	 */
	public function getPostUri()
	{
		return $this->post_uri;
	}

	/**
	 *  Get Response
	 *
	 *  Returns the entire response from PayPal as a string including all the
	 *  HTTP headers.
	 *
	 *  @return string
	 */
	public function getResponse()
	{
		return $this->response;
	}

	/**
	 *  Get Response Status
	 *
	 *  Returns the HTTP response status code from PayPal. This should be "200"
	 *  if the post back was successful.
	 *
	 *  @return string
	 */
	public function getResponseStatus()
	{
		return $this->response_status;
	}

	/**
	 *  Get Text Report
	 *
	 *  Returns a report of the IPN transaction in plain text format.
	 *  This is useful in emails to order processors and system administrators.
	 *
	 *  @return string
	 */
	public function getTextReport()
	{
		$r = '';
		$nl = PHP_EOL;
		$line = str_repeat('-', 80);

		// Current date and POST url
		$r .= $line.$nl;
		$r .= date('r').$nl;
		$r .= $this->getPostUri().$nl;

		// Used options
		$options = array();
		$options[] = ($this->use_curl ? 'curl' : 'fsockopen');
		$options[] = ($this->use_sandbox ? 'sandbox' : 'live');
		$options[] = ($this->verify_ssl ? 'verify ssl' : 'no verify ssl');
		$options[] = ($this->follow_location ? 'follow' : 'no follow');
		$options[] = 'timeout '.$this->timeout;
		$r .= '('.implode(', ', $options).')'.$nl;

		// HTTP response
		$r .= $line.$nl;
		$r .= $this->getResponse().$nl;

		// IPN data
		$r .= $line.$nl;
		if (!is_array($this->postData))
			$r .= print_r($this->postData, true);
		else
			foreach ($this->postData as $key => $value)
				$r .= str_pad($key, 25).$value.$nl;
		$r .= $line.$nl;

		return $r;
	}

	/**
	 * Process an Instant Payment Notification (IPN).
	 * 
	 *   1. Load POST data or data supplied via <code>$custom_data</code>.
	 *   2. Generate and submit a validation request for PayPal.
	 *   3. Parse the response from PayPal.
	 * 
	 * Call this method from your IPN listener script.
	 *
	 * @param mixed $custom_data Custom supplied data in a form of
	 *   a query string or an associative array. If NULL, POST data is used.
	 *   Default is NULL.
	 * @return boolean Returns TRUE if the response came back as "VERIFIED",
	 *   FALSE if the response came back as "INVALID", and throws an exception
	 *   if there was an error.
	 * @throws Exception
	 */
	public function processIpn($custom_data=null)
	{
		// Load POST data (avoid use of $_POST superglobal)
		if ($custom_data === null)
		{
			if ($this->requirePostMethod)
				$this->requirePostMethod();
			$this->rawPostData = file_get_contents('php://input');
			$this->postData = $this->decodeQueryString($this->rawPostData);
		}
		// Load custom data as an associative array
		else if (is_array($custom_data))
		{
			$this->postData = $custom_data;
			$this->rawPostData = $this->encodeQueryString($this->postData);
		}
		// Load custom data as a query string
		else
		{
			$this->rawPostData = $custom_data;
			$this->postData = $this->decodeQueryString($this->rawPostData);
		}

		// Generate request data to be sent to PayPal. Use the original
		// raw data string to avoid potential validation failure due to
		// encoding/decoding differences between client and server.
		$request_qeury = 'cmd=_notify-validate'.'&'.$this->rawPostData;

		// Submit request to PayPal and get response.
		if ($this->use_curl)
			$response = $this->curlPost($request_qeury);
		else
			$response = $this->fsockPost($request_qeury);

		// Check response HTTP status code
		if (strpos($response, '200') === false)
			throw new Exception("Invalid response status: " . $response);

		// Split response headers and payload, a better way for strcmp
		$tokens = explode("\r\n\r\n", trim($response));
		$response = trim(end($tokens));
		if (strcmp ($response, "VERIFIED") == 0)
			return true;
		else if (strcmp ($response, "INVALID") == 0)
			return false;
		else
			throw new Exception("Unexpected response from PayPal.");
	}

	/**
	 * Process IPN by calling <code>processIpn()</code> but catch the exception
	 * if error occurs and return it in the <code>$error</code> parameter.
	 *
	 * @param mixed $custom_data
	 * @param string $error Populated with an error message.
	 * @return boolean
	 */
	public function tryProcessIpn($custom_data=null, &$error=null)
	{
		try
		{
			$valid = $this->processIpn($custom_data);
			if (!$valid)
				$error = 'Invalid IPN request.';
			return $valid;
		}
		catch (Exception $e)
		{
			$error = $e->getMessage();
			return false;
		}
	}

	/**
	 *  Require Post Method
	 *
	 *  Throws an exception and sets a HTTP 405 response header if the request
	 *  method was not POST.
	 */
	public function requirePostMethod()
	{
		// Cannot use filter_input(INPUT_SERVER) due to a buggy behaviour of PHP.
		$method = isset($_SERVER['REQUEST_METHOD']) ? $_SERVER['REQUEST_METHOD'] : null;
		if ($method != 'POST')
		{
			header('Allow: POST', true, 405);
			throw new Exception("Invalid HTTP request method.");
		}
	}

	/**
	 * Encode an associative array into a query string.
	 * Optionally, add a question mark (?) at the beginning of the query string.
	 * @param array $data
	 * @param boolean $addQuestionMark
	 * @return string
	 */
	public static function encodeQueryString(array $data, $addQuestionMark=false)
	{
		$query = http_build_query($data);
		if ($addQuestionMark)
			$query = '?'.$query;
		return $query;
	}

	/**
	 * Decode a query string into an associative array with URL-decoded values.
	 * Returned data mymics the structure of <code>$_GET</code> and
	 * <code>$_POST</code> superglobals. Function automatically strips the
	 * question mark (?) at the beginning of the query string.
	 * @param string $query Query string.
	 * @return array
	 */
	public static function decodeQueryString($query)
	{
		// Strip '?' sign at the begining, if it exists
		if (strlen($query) > 0)
			if (substr($query, 0, 1) == '?')
				$query = substr($query, 1);

		// Break down query string into associative array
		$data = array();
		$parts = explode('&', $query);
		foreach ($parts as $keyval)
		{
			$keyval = explode('=', $keyval);
			if (count($keyval) == 2)
			{
				list($key, $value) = $keyval;
				$data[$key] = urldecode($value);
			}
		}
		
		return $data;
	}
}


examples/ipn_log.php
<?php

/*
 * This script verifies and logs to a file all received IPN requests.
 * You can use it with PayPal IPN Simulator, just upload it to your server
 * and specify its path as IPN handler URL:
 * https://developer.paypal.com/developer/ipnSimulator/
 *
 * Make sure that your web server has write access to the folder with this script,
 * otherwise you won't see any logs being created.
 */

include(dirname(dirname(__FILE__)).'/src/IpnListener.php');

use dezlov\PayPal\IpnListener;

$listener = new IpnListener();
$listener->use_sandbox = true;

$error = null;
$verified = $listener->tryProcessIpn(null, $error);
$report = $listener->getTextReport();
ipn_log($verified, $report, $error);

function ipn_log($verified, $report, $error)
{
	if ($verified)
	{
		$filename = 'ipn_verified.log';
		$content = $report;
	}
	else
	{
		$filename = 'ipn_errors.log';
		$content = 'ERROR: '.$error.PHP_EOL.$report;
	}
	file_put_contents($filename, $content, FILE_APPEND | LOCK_EX);
}


examples/ipn_basic_exceptions.php
<?php

/*
 * A basic script which will verify IPN via the sandbox server.
 * If an error occurs inside of processIpn() we will catch it
 * and output full text report.
 *
 * Note that 'echo' commands are just for demonstration purposes.
 * In reality, your script is going to be called by PayPal server and you
 * will not actually see the output unless you log it somewhere on your server.
 */

include(dirname(dirname(__FILE__)).'/src/IpnListener.php');

use dezlov\PayPal\IpnListener;

$listener = new IpnListener();
$listener->use_sandbox = true;

try
{
	$verified = $listener->processIpn();
	if ($verified)
		echo 'IPN request was "VERIFIED".'.PHP_EOL;
	else
		echo 'IPN request was "INVALID".'.PHP_EOL;
}
catch (Exception $e)
{
	echo 'IPN error: '.$e->getMessage().PHP_EOL;
	echo $listener->getTextReport();
}


examples/ipn_advanced.php
<?php

/*
 * Advanced example which demonstrates the complete processing chain of IPN,
 * including validation of IPN, verification of transaction details,
 * order fulfilment and error logging.
 *
 * This example follows the recommended practices of IPN integration from PayPal:
 * https://developer.paypal.com/docs/classic/ipn/integration-guide/IPNIntro/
 *
 * Dummy functions at the bottom of the script give examples of how to actually
 * handle various parts of the process.
 */

include(dirname(dirname(__FILE__)).'/src/IpnListener.php');

use dezlov\PayPal\IpnListener;

$listener = new IpnListener();
$listener->use_sandbox = true;

try
{
	// Process IPN request
	$verified = $listener->processIpn();
	if (!$verified)
		throw new Exception('Invalid IPN request');

	// Skip all notifications except for completed payments
	$payment_status = $listener->getData('payment_status');
	if ($payment_status != 'Completed')
		exit;

	// Check transaction ID
	$transaction_id = $listener->getData('txn_id');
	if (!check_transaction($transaction_id))
		throw new Exception('Unable to process IPN for transaction ID: '.$transaction_id);

	// Check receiver email
	$receiver_email = $listener->getData('receiver_email');
	if (!check_receiver_email($receiver_email))
		throw new Exception('Unable to process IPN for receiver email: '.$receiver_email);

	// Check price and currency
	$payment_total = $listener->getData('mc_gross');
	$payment_currency = $listener->getData('mc_currency');
	if (!check_price($payment_total, $payment_currency))
		throw new Exception('Unable to process IPN due to issues with price/currency');

	// If we got this far, then its ok to fulfil the order
	$item_name = $listener->getData('item_name');
	$item_number = $listener->getData('item_number');
	$payer_name = trim($listener->getData('first_name').' '.$listener->getData('last_name'));
	$payer_email = $listener->getData('payer_email');
	process_order($transaction_id, $item_name, $item_number, $payer_name, $payer_email);

	// Tell PayPal that we have successfully processing IPN.
	header('HTTP/1.1 200 OK');
}
catch (Exception $e)
{
	// Tell PayPal that we had problems processing IPN.
	header('HTTP/1.1 500 Internal Server Error');

	// Report error message.
	$message = 'IPN error: '.((string)$e);
	$message .= PHP_EOL.PHP_EOL.$listener->getTextReport();
	report_problem($message);
}



function process_order($transaction_id, $item_name, $item_number, $payer_name, $payer_email)
{
	// For example: Fulfil the order, save details in the database.
}

function check_price($total, $currency)
{
	// For example: Check that payment matches your advertised price.
	return true;
}

function check_receiver_email($email)
{
	// For example: Check that email matches your PayPal account.
	return true;
}

function check_transaction($transaction_id)
{
	// For example: Check that transaction ID does not yet exist if your database.
	return true;
}

function report_problem($message)
{
	// For example: Log via PHP error log.
	error_log($message);
}
